src/group_theory/perm/option.lean

```diff
@@ -0,0 +1,84 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Eric Wieser
+-/
+import group_theory.perm.basic
+import data.equiv.option
+import data.equiv.fin
+import data.fintype.basic
+
+/-!
+# Permutations of `option α`
+-/
+open equiv
+
+lemma equiv_functor.map_equiv_option_injective {α β : Type*} :
+  function.injective (equiv_functor.map_equiv option : α ≃ β → option α ≃ option β) :=
+equiv_functor.map_equiv.injective option option.some_injective
+
+@[simp]
+lemma map_equiv_remove_none {α : Type*} [decidable_eq α] (σ : perm (option α)) :
+  equiv_functor.map_equiv option (remove_none σ) = swap none (σ none) * σ :=
+begin
+  ext1 x,
+  have : option.map ⇑(remove_none σ) x = (swap none (σ none)) (σ x),
+  { cases x,
+    { simp },
+    { cases h : σ (some x),
+      { simp [remove_none_none _ h], },
+      { have hn : σ (some x) ≠ none := by simp [h],
+        have hσn : σ (some x) ≠ σ none := σ.injective.ne (by simp),
+        simp [remove_none_some _ ⟨_, h⟩, ←h, swap_apply_of_ne_of_ne hn hσn] } } },
+  simpa using this,
+end
+
+/-- Permutations of `option α` are equivalent to fixing an
+`option α` and permuting the remaining with a `perm α`.
+The fixed `option α` is swapped with `none`. -/
+@[simps] def equiv.perm.decompose_option {α : Type*} [decidable_eq α] :
+  perm (option α) ≃ option α × perm α :=
+{ to_fun := λ σ, (σ none, remove_none σ),
+  inv_fun := λ i, swap none i.1 * (equiv_functor.map_equiv option i.2),
+  left_inv := λ σ, by simp,
+  right_inv := λ ⟨x, σ⟩, begin
+    have : remove_none (swap none x * equiv_functor.map_equiv option σ) = σ :=
+      equiv_functor.map_equiv_option_injective (by simp [←mul_assoc, equiv_functor.map]),
+    simp [←perm.eq_inv_iff_eq, equiv_functor.map, this],
+  end }
+
+/-- Permutations of `fin (n + 1)` are equivalent to fixing a single
+`fin (n + 1)` and permuting the remaining with a `perm (fin n)`.
+The fixed `fin (n + 1)` is swapped with `0`. -/
+def equiv.perm.decompose_fin {n : ℕ} :
+  perm (fin n.succ) ≃ fin n.succ × perm (fin n) :=
+((equiv.perm_congr $ fin_succ_equiv n).trans equiv.perm.decompose_option).trans
+  (equiv.prod_congr (fin_succ_equiv n).symm (equiv.refl _))
+
+@[simp] lemma equiv.perm.decompose_fin_symm_of_refl {n : ℕ} (p : fin (n + 1)) :
+  equiv.perm.decompose_fin.symm (p, equiv.refl _) = swap 0 p :=
+begin
+  ext x,
+  by_cases hp : p = 0;
+  by_cases hx : x = 0;
```

Comment:
Does the proof become easier if you use `cases hp : p` instead of `cases hp : p = 0` here ad similarly for `x`?

Reply:
In this case, it's a bit more complicated because `x` and `p` are both `fin (n + 1)` so a simple `cases` just deconstructs the subtype into the value and property witness. Using `refine fin.cases _ _ x` also is not great because the `hx' : x = p` or the inequality is "forgotten" -- the second goal does not relate the now universally quantified `i : fin n` back to `x`.

Reply:
But I think there are some ways to shorten it -- working on it

Reply:
```
@[simp] lemma equiv.perm.decompose_fin_symm_of_refl {n : ℕ} (p : fin (n + 1)) :
  equiv.perm.decompose_fin.symm (p, equiv.refl _) = swap 0 p :=
begin
  ext x,
  refine fin.cases _ _ p;
  refine fin.cases _ _ x;
  simp [equiv.perm.decompose_fin],
  intros x₁ p₁,
  by_cases h : x₁ = p₁;
  simp [h, equiv.perm.decompose_fin, swap_apply_of_ne_of_ne, fin.succ_ne_zero]
end
```
Is that preferred?

Reply:
That looks reasonable for me - I'm waiting for CI to generate a cache for me so that I can step through it.

Reply:
Having played around with this - ideally we can get rid of the `ext`, and have simp handle it all. This lemma helps:
```lean
@[simp] lemma map_equiv_refl {α : Type*} : equiv_functor.map_equiv option (1 : perm α) = 1 := begin
  ext, simp [equiv_functor.map_equiv, equiv_functor.map],
end
```
as does normalizing with `simp [←equiv.perm.one_def, ←equiv.perm.inv_def]`

Reply:
I've gone ahead and pushed a commit with these new lemmas, which reduces this proof to `simp [equiv.perm.decompose_fin]`.